p2p/enr/enr.go

```diff
@@ -0,0 +1,293 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package discover implements the Ethereum Node Record as per https://github.com/ethereum/EIPs/pull/778
+package enr
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"errors"
+	"io"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/rlp"
+
+	"github.com/btcsuite/btcd/btcec"
+)
+
+var ()
+
+// The maximum encoded size of a node record is 300 bytes. Implementations should reject records larger than this size.
+const (
+	ID_SECP256k1_KECCAK = "secp256k1-keccak" // "secp256k1-keccak" identity scheme identifier
+)
+
+// Key is implemented by known node record key types.
+//
+// To define a new key that is to be included in a node record,
+// create a Go type that satisfies this interface. The type should
+// also implement rlp.Decoder if additional checks are needed on the value.
+type Key interface {
+	ENRKey() string
+}
+
+type pair struct {
+	k []byte
```

Comment:
Package rlp supports strings. Please make `k` a string to avoid conversions.

Reply:
Done.